From 31a39bc277bbbcbc30748689603833beb86534f1 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Fri, 26 Mar 2010 12:06:50 +0800
Subject: [PATCH] CN50xx Kexec: fix the kernel panic when the kernel reboot bypassing the bootloader

This issue is caused by the second kernel can not get usable memory from firmware.

This board has only one DIMM slot. And u-boot can only use the 512MB memory.
delta ( memory(512MB) - freemem ) is also less than 512MB(MAX_MEMORY) when the
second kernel start from the first kernel. So the second kernel can not call the
__cvmx_bootmem_phy_free() to release the used memory.

We set a free memory threshold (256MB). It need to release the memory when the free
memory less than 256MB.

Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 arch/mips/cavium-octeon/setup.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/mips/cavium-octeon/setup.c b/arch/mips/cavium-octeon/setup.c
index bc69f3c..6c95935 100644
--- a/arch/mips/cavium-octeon/setup.c
+++ b/arch/mips/cavium-octeon/setup.c
@@ -329,7 +329,7 @@ void cleanup_phy_mem(void)
 	freemem = cvmx_bootmem_phy_available_mem(0x100000);
 	delta = (memory > freemem ? (memory - freemem) : (freemem - memory));
 
-	if (delta < MAX_MEMORY) {
+	if ((delta < MAX_MEMORY) && (freemem >= 0x10000000)) {
 		/* we haven't allocated any memory yet = no need to
 		 * continue free'ing
 		 */
@@ -347,7 +347,7 @@ void cleanup_phy_mem(void)
 	curr_addr += 1;
 	curr_addr = curr_addr << 20;
 
-	memory -= kernel_size;
+	memory -= curr_addr;
 
 	if (memory <= (OCTEON_DDR0_SIZE - curr_addr)) {
 		__cvmx_bootmem_phy_free(curr_addr, memory, 0);
-- 
1.6.0.4

